controllers/deployment.go

```diff
@@ -343,14 +354,56 @@ func buildDeployment(master *v1alpha1.CDAPMaster, name string, services ServiceG
 	return spec, nil
 }
 
+type ByEnvKey []corev1.EnvVar
+
+func (k ByEnvKey) Len() int           { return len(k) }
+func (k ByEnvKey) Swap(i, j int)      { k[i], k[j] = k[j], k[i] }
+func (k ByEnvKey) Less(i, j int) bool { return k[i].Name < k[j].Name }
+
+func mergeEnvVars(baseEnvVars []corev1.EnvVar, overwriteEnvVars []corev1.EnvVar) ([]corev1.EnvVar, error) {
+	// Merge master and service-specific environment variables
+	envMap := make(map[string]corev1.EnvVar)
+	// Add master environment variables.
+	for _, baseEnvVar := range baseEnvVars {
+		if _, ok := envMap[baseEnvVar.Name]; ok {
+			return nil, fmt.Errorf("duplicate env var %q in base slice", baseEnvVar.Name)
+		}
+		envMap[baseEnvVar.Name] = baseEnvVar
+	}
+
+	// Add and overwrite the service-specific environment variables.
+	// Maintain a seen map and throw an error if there are duplicates in the service env var slice.
+	seenVars := make(map[string]bool)
+	for _, envVar := range overwriteEnvVars {
+		if _, ok := seenVars[envVar.Name]; ok {
+			return nil, fmt.Errorf("duplicate env var %q in overwrite slice", envVar.Name)
+		}
+		seenVars[envVar.Name] = true
+		envMap[envVar.Name] = envVar
+	}
+
+	// Convert the map to a sorted slice.
+	env := []corev1.EnvVar{}
+	for _, envVar := range envMap {
+		env = append(env, envVar)
+	}
+	sort.Sort(ByEnvKey(env))
```

Comment:
nit: Why do we need to sort?

Reply:
Consistent ordering seems like it'll be nicer to have (to spot errors, etc), and we could end up with flaky tests if we don't sort as well. Performance loss should be minimal since we're unlikely to be adding a huge amount of environment variables, but if it does become a problem, we can also just move the sort function to the test itself.